File: dsa-problems/Probabilistic/solutions/java/PRB-002-expected-steps-random-walk-1d.java
import java.util.*;

class Solution {
    public double expectedSteps(int a, int b, double p) {
        if (Math.abs(p - 0.5) < 1e-9) {
            return (double) a * b;
        }

        double q = 1.0 - p;
        double r = q / p;
        double M = a + b;
        double z = b;

        double term1 = z / (q - p);
        double term2 = (M / (q - p)) * ((1.0 - Math.pow(r, z)) / (1.0 - Math.pow(r, M)));

        return term1 - term2;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (sc.hasNextInt()) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            double p = sc.nextDouble();
            Solution solution = new Solution();
            System.out.println(solution.expectedSteps(a, b, p));
        }
        sc.close();
    }
}

File: dsa-problems/Probabilistic/solutions/cpp/PRB-002-expected-steps-random-walk-1d.cpp
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

class Solution {
public:
    double expectedSteps(int a, int b, double p) {
        if (abs(p - 0.5) < 1e-9) {
            return (double)a * b;
        }

        double q = 1.0 - p;
        double r = q / p;
        double M = a + b;
        double z = b;

        double term1 = z / (q - p);
        double term2 = (M / (q - p)) * ((1.0 - pow(r, z)) / (1.0 - pow(r, M)));

        return term1 - term2;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int a, b;
    double p;
    if (cin >> a >> b >> p) {
        Solution solution;
        cout << fixed << setprecision(6) << solution.expectedSteps(a, b, p) << "\n";
    }
    return 0;
}
